Source/NSTextView.m

```diff
@@ -6078,6 +6081,88 @@ - (void) performFindPanelAction: (id)sender
 		withTextView: self];
 }
 
+// NSTextFinder methods implementation...
+// isSelectable, isEditable, string, selectedRanges, setSelectedRanges, replaceCharactersInRange:withString:
+// implemented by NSTextView already...
+
+- (BOOL) allowsMultipleSelection
+{
+  return NO;
+}
+
+- (NSString *) stringAtIndex: (NSUInteger)characterIndex
+              effectiveRange: (NSRangePointer)outRange
+      endsWithSearchBoundary: (BOOL *)outFlag
+{
+  return nil;
+}
+
+- (NSUInteger) stringLength
+{
+  return [[self string] length];
+}
+
+- (NSRange) firstSelectedRange
+{
+  NSValue *r = [[self selectedRanges] objectAtIndex: 0];
+  return [r rangeValue];
+}
+
+- (BOOL) shouldReplaceCharactersInRanges: (NSArray *)ranges withStrings: (NSArray *)strings
+{
+  return YES;
+}
+
+- (void) didReplaceCharacters
+{
+  // nothing...
+}
+
+- (NSView *) contentViewAtIndex: (NSUInteger)index effectiveCharacterRange: (NSRangePointer)outRange
+{
+  return nil;
+}
+
+- (NSArray *) rectsForCharacterRange: (NSRange)range
+{
+  NSValue *value = [NSValue valueWithRect: [self rectForCharacterRange: range]];
+  NSArray *result = [NSArray arrayWithObject: value];
+  return result;
+}
+
+- (NSArray *) visibleCharacterRanges
+{
+  NSArray *result = nil;
+  
+  if (_layoutManager)
+    {
+      const NSRect visibleRect = [self visibleRect];
+      
+      NSRange visibleGlyphRange = [_layoutManager glyphRangeForBoundingRect: visibleRect
+                                                            inTextContainer: _textContainer];
+      
+      NSRange visibleRange = [_layoutManager characterRangeForGlyphRange: visibleGlyphRange
+                                                        actualGlyphRange: NULL];
+
+      NSValue *value = [NSValue valueWithRange: visibleRange];
+      result = [NSArray arrayWithObject: value];
+    }
+ 
+  return result;
+}
+
+- (void) drawCharactersInRange: (NSRange)range forContentView: (NSView *)view
+{
+  NSArray *rectsArray = [self rectsForCharacterRange: range];
```

Comment:
I think the code here should rather contain the two calls to the layout manager from the method drawRect:

Reply:
Please elaborate, if possible.

Reply:
```
  /* Then draw the special background of the new glyphs.  */
  [_layoutManager drawBackgroundForGlyphRange: drawnRange
                  atPoint: _textContainerOrigin];

  [_layoutManager drawGlyphsForGlyphRange: drawnRange 
                  atPoint: _textContainerOrigin];
```
Do you mean these?

Reply:
On the assumption that that is what you meant, done.